Feature/adding location

delcroip opened the pull request.

weilu committed:
Add village to individual and group model

to facilitate row security

weilu committed:
Unit test existing individual GQL permission control

weilu committed:
Enable row level security on group and individual list queries

weilu committed:
Row-level security for individual history query

weilu committed:
Row-level security for group history query

weilu committed:
Row-level security for group membership query

Aka GroupIndividual

weilu committed:
Row-level security for individual group membership history

weilu committed:
Add row level security to individual creation query

weilu committed:
Add row-level security to individual update query

weilu committed:
Add row-level security to individual delete query

weilu committed:
Add row-level security to individual undo delete query

weilu committed:
Add basic create group gql mutation test

Also refactor individual mutation & query test to extract setup code

weilu committed:
Add row level security to group create

weilu committed:
Row level security for group update

weilu committed:
Add row level security to group delete

weilu committed:
Add row level security to adding individual to group

weilu committed:
Verify that both individual_id and group_id are required by service

weilu committed:
Add row level security to editing individual in group

weilu committed:
Add row level security to removing individuals from group

weilu committed:
Make sure individual group move doesn’t create and delete unnecessarily

dborowiecki committed:
Fix perforamnce for individual data update (#127)

* Fix perforamnce for individual data update

* added saving in bulk way validation errors

* removed redundant print

---------

Co-authored-by: sniedzielski <sniedzielski@soldevelo.com>

weilu committed:
Remove unnecessary location-individual relationship reversal in query

delcroip committed:
Merge remote-tracking branch 'wei/row-security' into release/24.10

delcroip committed:
fixing tests

delcroip commented on individual/tests/graphql_mutation_group_test.py:
```diff
@@ -722,17 +732,17 @@ def test_remove_individuals_from_group_row_security(self):
-        content = json.loads(response.content)
-        id = content['data']['removeIndividualFromGroup']['internalId']
-        self.assert_mutation_success(id)
+        # group_no_loc = create_group(self.admin_user.username)
```
@weilu this test create a wierd error Group is not callable, please have a look I have sick of this module after fixing test for several days

weilu replied:
I uncommented the test case and it passed for me locally with the location module pointed to the `feature/is-allowed` branch. Can you paste the full stacktrace?

weilu committed:
Fix create group and move individual service test

delcroip committed:
Merge branch 'develop' into feature/adding-location

weilu committed:
Put back previously commented out test case

weilu commented on individual/tests/graphql_mutation_group_test.py:
```diff
@@ -613,11 +624,7 @@ def test_edit_individual_in_group_row_security(self):
         )
         content = json.loads(response.content)
-        id = content['data']['editIndividualInGroup']['internalId']
-        self.assert_mutation_error(id, 'mutation.individual_group_village_mismatch')
```
@delcroip This test case shouldn't have been deleted as it catches the case that an imis admin user can manipulate any individual's group assignment but we should validate to ensure the individual and their assigned group's locations are consistent.

weilu commented on individual/tests/graphql_mutation_group_test.py:
```diff
@@ -175,24 +180,27 @@ def test_update_group_row_security(self):
 
         # SP officer B cannot update group for district A
         response = self.query(query_str)
+        content = json.loads(response.content)
+        internal_id = content['data']['updateGroup']['internalId']
+        self.assert_mutation_error(internal_id, 'mutation.authentication_required')
```
This shouldn't be necessary as the anonymous user case is already tested in the test above. I see why you added them – probably due to the extraneous line I have above where the response was not checked. I'll push a change to remove this and similar changes below.

delcroip replied:
I added it to force the execution of the mutation, if not error could happen later in the process, if we remove the get_mutation_result we should also remove the query

weilu replied:
Yes, I removed the query in https://github.com/openimis/openimis-be-individual_py/pull/129/commits/324c578daf909a9b374d37d20ed1d8b04907a4db

weilu reviewed (changes requested):
@delcroip The change to use `LocationManager.is_allowed` instead of `Location.get_queryset` introduces the following issues: 
1. it no longer considers and respects `settings.ROW_SECURITY`,
2.  users with `user.is_imis_admin = True` now is not permitted to mutate any individual or group with location. Case 2 is actually covered by a test case you removed (see my comment in code).

In other words, `Location.get_queryset` offers more functionalities needed than that `LocationManager.is_allowed` is currently offering. In the spirit of DRY, I'd suggest not replicating the checks inside `is_allowed`, but considering pushing the caching logic into `LocationManager.allowed` which is used by `Location.get_queryset` and keep using `Location.get_queryset` for location authorization checking.

weilu committed:
Remove repeated tests & add back test case on group-individual loc match

- Also modify existing cases to ensure imis admin can modify individual
and group with location
- Show error message when assert_mutation_success fails

delcroip committed:
Merge remote-tracking branch 'origin/develop' into feature/adding-location

delcroip committed:
Merge branch 'feature/adding-location' of https://github.com/openimis/openimis-be-individual_py into feature/adding-location

delcroip committed:
manage not existing individualGroup

weilu commented on individual/tests/graphql_mutation_group_test.py:
```diff
@@ -474,7 +474,7 @@ def test_add_individual_to_group_row_security(self):
         content = json.loads(response.content)
         internal_id = content['data']['addIndividualToGroup']['internalId']
-        self.assert_mutation_error(internal_id, _('unauthorized.location'))
+        self.assert_mutation_success(internal_id)
```
@delcroip why has this been changed to success? Shouldn't we prevent such a scenario to keep location data consistent between individual and their group?

weilu committed:
Fix no GroupIndividual found error in test

weilu committed:
Check that individual and group location match

On group assignments and updates
